kona/src/main.rs

```diff
@@ -0,0 +1,283 @@
+use std::{collections::HashMap, sync::Arc};
+
+use clap::Parser;
+use eyre::{bail, Result};
+use kona_derive::{
+    online::{AlloyL2ChainProvider, EthereumDataSource},
+    traits::{ChainProvider, L2ChainProvider, OriginProvider, Pipeline, StepResult},
+    types::{BlockInfo, L2BlockInfo, RollupConfig, StageError, OP_MAINNET_CONFIG},
+};
+use reth::{cli::Cli, transaction_pool::TransactionPool};
+use reth_exex::{ExExContext, ExExEvent, ExExNotification};
+use reth_node_api::FullNodeComponents;
+use reth_node_ethereum::EthereumNode;
+use tracing::{debug, error, info, warn};
+
+mod blobs;
+use blobs::ExExBlobProvider;
+
+mod cli_ext;
+use cli_ext::{KonaArgsExt, ValidationMode};
+
+mod pipeline;
+use pipeline::{new_local_pipeline, LocalAttributesBuilder, LocalPipeline};
+
+mod providers;
+use providers::LocalChainProvider;
+
+mod validation;
+use validation::{AttributesValidator, EngineApiValidator, TrustedValidator};
+
+#[derive(Debug)]
+pub(crate) struct KonaExEx<Node: FullNodeComponents> {
+    /// The rollup configuration
+    cfg: Arc<RollupConfig>,
+    /// The context of the Execution Extension
+    ctx: ExExContext<Node>,
+    /// The chain provider to follow the chain state based on ExEx events
+    chain_provider: LocalChainProvider,
+    /// The L2 chain provider to fetch L2 block information and optionally
+    /// verify newly derived payloads against
+    l2_provider: AlloyL2ChainProvider,
+    /// The blob provider to fetch blobs from the beacon client
+    blob_provider: ExExBlobProvider,
+    /// The validator to verify newly derived payloads
+    validator: Box<dyn AttributesValidator + Send>,
```

Comment:
Could also make this a generic param on the `KonaExEx` struct, either way should work.

Reply:
I opted for this instead of having to specify the type when creating the ExEx in `main` fn. I can do the generic if you think it's cleaner (all my homies hate dynamic dispatch)